### PR TITLE
Re-introduce PR#1030 baseDomain patch and add backend logic

### PR DESCRIPTION
**Pull request by mangelajo**
- Re-introduce PR#1030 baseDomain patch
- [USHIFT-664](https://issues.redhat.com//browse/USHIFT-664) Backend logic to baseDomain

### CODE REVIEW

**Comment by mangelajo** [comment]
@fzdarsky @dinhxuanvu , please see the two commit structure, on first commit, I introduce @dinhxuanvu original patch, on second commit I handle the logic to match OpenShift behavior, also introducing the fixed "cluster.local" setting.

File: pkg/controllers/kube-apiserver.go [hunk 1]
```diff
@@ -165,7 +165,7 @@ func (s *KubeAPIServer) configure(cfg *config.MicroshiftConfig) error {
 									APIVersion: "route.openshift.io/v1",
 									Kind:       "HostAssignmentAdmissionConfig",
 								},
-								Domain: cfg.Cluster.Domain,
+								Domain: cfg.BaseDomain, // @benluddy, would this be the right thing?
```
Comment (mangelajo) [inline pkg/controllers/kube-apiserver.go:168]: @benluddy ^ would this be the right thing ?

File: docs/howto_config.md [hunk 2]
```diff
@@ -33,7 +34,7 @@ The configuration settings alongside with the supported command line arguments a
 | cidr (clusterNetwork) | --cluster-cidr            | MICROSHIFT_CLUSTER_CLUSTERCIDR          | A block of IP addresses from which Pod IP addresses are allocated
 | serviceNetwork        | --service-cidr            | MICROSHIFT_CLUSTER_SERVICECIDR          | A block of virtual IP addresses for Kubernetes services
 | serviceNodePortRange  | --service-node-port-range | MICROSHIFT_CLUSTER_SERVICENODEPORTRANGE | The port range allowed for Kubernetes services of type NodePort
-| clusterDomain         | --cluster-domain          | MICROSHIFT_CLUSTER_DOMAIN               | Base DNS domain used to construct fully qualified pod and service domain names
+| baseDomain            | --cluster-domain          | MICROSHIFT_CLUSTER_DOMAIN               | Base DNS domain used to construct fully qualified pod and service domain names
```
Comment (mangelajo) [inline docs/howto_config.md:37]:
```suggestion
| baseDomain            | --cluster-domain          | MICROSHIFT_BASEDOMAIN                 | Base DNS domain used to construct fully qualified pod and service domain names
```

File: pkg/controllers/kube-apiserver.go [hunk 3]
```diff
@@ -165,7 +165,7 @@ func (s *KubeAPIServer) configure(cfg *config.MicroshiftConfig) error {
 									APIVersion: "route.openshift.io/v1",
 									Kind:       "HostAssignmentAdmissionConfig",
 								},
-								Domain: cfg.Cluster.Domain,
+								Domain: cfg.BaseDomain,
```
Comment (mangelajo) [inline pkg/controllers/kube-apiserver.go:168]: @benluddy would this be the right thing to use here ^ ?  before we were using cluster.local but I wonder if this is more related to  https://github.com/openshift/microshift/pull/1176/files#diff-c70ba87d9ac52efa40c2eb55913d76e6c263c9690a9f3e48b074865cbd366738L65  ?

File: pkg/cmd/init.go [hunk 4]
```diff
@@ -214,7 +214,8 @@ func certSetup(cfg *config.MicroshiftConfig) (*certchains.CertificateChains, err
 					ValidityDays: cryptomaterial.ShortLivedCertificateValidityDays,
 				},
 				Hostnames: []string{
-					"router-default.apps." + cfg.Cluster.Domain,
+					"router-default.apps." + cfg.BaseDomain,
+					"*.apps." + cfg.BaseDomain, // wildcard for any additional auto-generated domains
```
Comment (mangelajo) [inline pkg/cmd/init.go:218]: I checked on an full OpenShift:

[majopela@rhel9 microshift]$ openssl s_client -connect oauth-openshift.apps.ci-ln-n3z62kb-76ef8.aws-2.ci.openshift.org:443  </dev/null 2>/dev/null | openssl x509 -noout -text | grep DNS
                DNS:*.apps.ci-ln-n3z62kb-76ef8.aws-2.ci.openshift.org

Probably we can remove 217 as it is redundant,

File: docs/howto_config.md [hunk 5]
```diff
@@ -48,13 +49,14 @@ In case `config.yaml` is not provided, the following default settings will be us
...
 url: https://127.0.0.1:6443
-clusterDomain: cluster.local
 network:
   clusterNetwork:
     - cidr: 10.42.0.0/16
   serviceNetwork:
     - 10.43.0.0/16
   serviceNodePortRange: 30000-32767
+dns:
+  baseDomain: cluster.local
```
[the remark this answers is not in the document]
Comment (dhellmann) [reply docs/howto_config.md:59]: Maybe the example should be `microshift.example.com`?

File: pkg/cmd/init.go (hunk 4 above)
Comment (dhellmann) [reply pkg/cmd/init.go:218]: @stlaz and @oglok are our cert experts, let's get them to confirm.

File: pkg/cmd/init.go [hunk 6]
```diff
@@ -271,6 +278,7 @@ func certSetup(cfg *config.MicroshiftConfig) (*certchains.CertificateChains, err
 					"openshift.default",
 					"openshift.default.svc",
 					"openshift.default.svc.cluster.local",
+					"api-int." + cfg.BaseDomain, // TODO: OpenShift actually uses  api.$ClusterName.$BaseDomain
```
Comment (dhellmann) [inline pkg/cmd/init.go:281]: Don't we have both for OCP? Do we need both here?

File: pkg/config/config.go [hunk 7]
```diff
@@ -61,22 +60,23 @@ type IngressConfig struct {
...
 
-// Top level config
+// Top level config file
 type Config struct {
 	NodeName        string    `json:"nodeName"`
 	NodeIP          string    `json:"nodeIP"`
 	URL             string    `json:"url"`
-	ClusterDomain   string    `json:"clusterDomain"`
 	Network         Network   `json:"network"`
+	ClusterName     string    `json:"clusterName"`
```
Comment (dhellmann) [inline pkg/config/config.go:78]: I don't see `ClusterName` being used in this PR. We had a discussion on slack about what the default for the cluster name should be, how to report it in the cluster, how to keep it from being changed, etc. so I think adding it will be a big enough change to deserve its own PR.

File: pkg/config/config.go [hunk 8]
```diff
@@ -106,6 +106,17 @@ type ClusterNetworkEntry struct {
...
 }
 
+type DNS struct {
+	// baseDomain is the base domain of the cluster. All managed DNS records will
+	// be sub-domains of this base.
+	//
+	// For example, given the base domain `openshift.example.com`, an API server
+	// DNS record may be created for `cluster-api.openshift.example.com`.
+	//
+	// Once set, this field cannot be changed.
```
Comment (dhellmann) [inline pkg/config/config.go:116]: Do we need another ticket to protect against the value changing?

**Review by dinhxuanvu — CHANGES_REQUESTED** [review]
Just a few minor nits. Overall, the PR looks good.

File: pkg/config/config.go (hunk 7 above)
Comment (dinhxuanvu) [reply pkg/config/config.go:78]: Just as Doug pointed out, I do not see the usage of this `ClusterName` field. Perhaps, this is unintentional. I recommend removing it given it is not relevant to the changes in this PR.

**Review by dinhxuanvu — CHANGES_REQUESTED** [review]

File: pkg/config/config.go (hunk 8 above)
Comment (dgrisonnet) [reply pkg/config/config.go:116]: I believe that this is a comment inherited from the OCP config API: https://github.com/openshift/api/blob/master/config/v1/types_dns.go#L27-L34

We might want to update it if it doesn't apply to MicroShift

File: pkg/config/config.go (hunk 7 above)
Comment (dgrisonnet) [reply pkg/config/config.go:78]: it seems to be needed in https://github.com/openshift/microshift/pull/1176/commits/5147e3fac8b87e1096d3fc0c51bc9788c003f87c#diff-a53aedd081647bc5a1bc5c1a969ad049f4ff08c9b4c5ba9408ea647488849234R281, but it is not used yet. I am also leaning towards adding that in a separate effort.

File: pkg/config/config.go (hunk 8 above)
Comment (fzdarsky) [reply pkg/config/config.go:116]: It does apply to MicroShift, too, I think @dhellmann 's concern was more where/how this is enforced.

File: pkg/config/config.go (hunk 7 above)
Comment (fzdarsky) [reply pkg/config/config.go:78]: We're not just replacing a name here, though. What used to be "ClusterDomain" will become "ClusterName.BaseDomain".

Considering this, is your suggestion then that BaseDomain should change its semantic between this PR and the "separate effort"? If not, the minimum we have to dl is to at least update all locations where this PR replaces ClusterDomain with BaseDomain to add a hardcoded ClusterName.

File: pkg/cmd/init.go (hunk 4 above)
Comment (mangelajo) [reply pkg/cmd/init.go:218]: no, SNO / OpenShift uses a wildcard cert *.apps.clustername.basedomain

that means that any other applications exposed under that domain will have a "valid" (self signed, but OK), certificate...

Comment (mangelajo) [reply pkg/cmd/init.go:218]: See my check above @fzdarsky :

```
[majopela@rhel9 microshift]$ openssl s_client -connect oauth-openshift.apps.ci-ln-n3z62kb-76ef8.aws-2.ci.openshift.org:443 </dev/null 2>/dev/null | openssl x509 -noout -text | grep DNS
DNS:*.apps.ci-ln-n3z62kb-76ef8.aws-2.ci.openshift.org
```

File: pkg/config/config.go (hunk 7 above)
Comment (mangelajo) [reply pkg/config/config.go:78]: Sorry it's unintentional, it was intended to be introduced on a PR after this one , which we needed to discuss.

File: pkg/cmd/init.go (hunk 6 above)
Comment (mangelajo) [reply pkg/cmd/init.go:281]: Let me add both. I thought api was for the externally serving cert, and api-int for the internally serving. But it would harm nobody.

File: pkg/config/config.go (hunk 7 above)
Comment (mangelajo) [reply pkg/config/config.go:78]: I am adding "microshift" for now as the clustername part, 
on a 2nd PR we can make that configurable.

File: pkg/config/config.go (hunk 8 above)
Comment (mangelajo) [reply pkg/config/config.go:116]: We may want to enforce against changes, I agree.

File: pkg/cmd/init.go (hunk 4 above)
Comment (fzdarsky) [reply pkg/cmd/init.go]: Sorry for not being clear: I was trying to point out the missing `cfg.ClusterName` part (i.e. `ci-ln-n3z62kb-76ef8`). You're right, of course, that this is not only for the `router-default.aps.` line but also for the `*.apps.` line.

File: pkg/controllers/kube-apiserver.go (hunk 1 above)
Comment (fzdarsky) [reply pkg/controllers/kube-apiserver.go]: Missing `cfg.ClusterName`

File: pkg/controllers/kube-apiserver.go (hunk 3 above)
Comment (fzdarsky) [reply pkg/controllers/kube-apiserver.go:168]: Missing `cfg.ClusterName`, I assume.

File: pkg/config/config.go [hunk 9]
```diff
@@ -61,22 +59,22 @@ type IngressConfig struct {
...
 
-// Top level config
+// Top level config file
 type Config struct {
 	NodeName        string    `json:"nodeName"`
 	NodeIP          string    `json:"nodeIP"`
 	URL             string    `json:"url"`
-	ClusterDomain   string    `json:"clusterDomain"`
 	Network         Network   `json:"network"`
+	DNS             DNS       `json:"dns"`
```
Comment (fzdarsky) [inline pkg/config/config.go:77]: Can we call this `cluster`, please? The `baseDomain` (and possibly also the `clusterName` or whatever we call it) are properties of the cluster itself and used in more places than just DNS (which suggests CoreDNS), e.g. Ingress.

File: pkg/config/config.go [hunk 10]
```diff
@@ -180,12 +189,12 @@ func NewMicroshiftConfig() *MicroshiftConfig {
 		SubjectAltNames: subjectAltNames,
 		NodeName:        nodeName,
 		NodeIP:          nodeIP,
+		BaseDomain:      "microshift.example.com",
```
Comment (fzdarsky) [inline pkg/config/config.go:192]: Shouldn't this be "example.com", as "microshift" is also the clusterName?
(i.e. we'd end up getting`*.apps.microshift.microshift.example.com`)

File: pkg/cmd/init.go [hunk 11]
```diff
@@ -214,7 +214,7 @@ func certSetup(cfg *config.MicroshiftConfig) (*certchains.CertificateChains, err
 					ValidityDays: cryptomaterial.ShortLivedCertificateValidityDays,
 				},
 				Hostnames: []string{
-					"router-default.apps." + cfg.Cluster.Domain,
+					"*.apps.microshift." + cfg.BaseDomain, // wildcard for any additional auto-generated domains
```
Comment (fzdarsky) [inline pkg/cmd/init.go:217]: Is this change from `router-default` to `*` wildcard intentional?

File: docs/howto_config.md [hunk 12]
```diff
@@ -33,7 +34,7 @@ The configuration settings alongside with the supported command line arguments a
 | cidr (clusterNetwork) | --cluster-cidr            | MICROSHIFT_CLUSTER_CLUSTERCIDR          | A block of IP addresses from which Pod IP addresses are allocated
 | serviceNetwork        | --service-cidr            | MICROSHIFT_CLUSTER_SERVICECIDR          | A block of virtual IP addresses for Kubernetes services
 | serviceNodePortRange  | --service-node-port-range | MICROSHIFT_CLUSTER_SERVICENODEPORTRANGE | The port range allowed for Kubernetes services of type NodePort
-| clusterDomain         | --cluster-domain          | MICROSHIFT_CLUSTER_DOMAIN               | Base DNS domain used to construct fully qualified pod and service domain names
+| baseDomain            | --base-domain             | MICROSHIFT_BASEDOMAIN                   | Base DNS domain used to construct fully qualified pod and service domain names
```
Comment (fzdarsky) [inline docs/howto_config.md:37]: Documentation string needs update, as its semantic changed. (maybe just: "Base domain of the cluster.")

File: pkg/cmd/init.go (hunk 11 above)
Comment (mangelajo) [reply pkg/cmd/init.go:217]: Yes, see my comments; it is what OCP does. Providing a wildcard domain in this certificate.

In this way both router-default and any other router generated domains will be covered by this serving cert.

Comment (mangelajo) [reply pkg/cmd/init.go:217]: It is true that may be it does not exactly belong in this PR, but I was trying to align with what OCP did regarding the basedomain/clustername on the certs, and that's what I found.

```
[majopela@rhel9 microshift]$ openssl s_client -connect oauth-openshift.apps.ci-ln-n3z62kb-76ef8.aws-2.ci.openshift.org:443 </dev/null 2>/dev/null | openssl x509 -noout -text | grep DNS
DNS:*.apps.ci-ln-n3z62kb-76ef8.aws-2.ci.openshift.org
```

File: pkg/config/config.go (hunk 9 above)
Comment (mangelajo) [reply pkg/config/config.go:77]: That was part of the alignment @dinhxuanvu tried to do regarding openshift-install config, so I added it as-is. I would need him to weigh in. To me, your suggestion makes sense but would like to hear his opinion.

File: pkg/config/config.go (hunk 10 above)
Comment (mangelajo) [reply pkg/config/config.go:192]: +1 true, at least until we resolve/allow adding a cluster name

File: pkg/controllers/kube-apiserver.go (hunk 3 above)
Comment (mangelajo) [reply pkg/controllers/kube-apiserver.go:168]: I believe it probably makes sense to add a constant "ClusterName" to the config module that we can use instead of just adding "microshift." everywhere. Then the intent will be clearer and also, it will be simpler to convert into cfg.ClusterName afterwards.

Comment (mangelajo) [reply pkg/controllers/kube-apiserver.go:168]: Or even apps. + ClusterName + BaseDomain? 

I am not sure of the intent of this admission controller config. @benluddy ^

**Comment by oglok** [comment]
Squash at least last two commits, so we avoid having two with the same title in history.

File: pkg/cmd/init.go (hunk 11 above)
Comment (fzdarsky) [reply pkg/cmd/init.go:217]: Thanks for investigating this. Makes sense to me then,

File: pkg/config/config.go (hunk 9 above)
Comment (fzdarsky) [reply pkg/config/config.go:78]: Ok, I'd suggest to get this merged and do a follow-up PR, as this is not immediately related.

**Comment by mangelajo** [comment]
> Squash at least last two commits, so we avoid having two with the same title in history.

I added the merge method "squash" for tide to handle.
